src/app/child-dev-project/attendance/model/activity-attendance.ts

```diff
@@ -0,0 +1,150 @@
+import { Note } from "../../notes/model/note";
+import {
+  AttendanceCounting,
+  AttendanceStatus,
+  DEFAULT_ATTENDANCE_TYPES,
+} from "./attendance-status";
+import { Entity } from "../../../core/entity/entity";
+import { RecurringActivity } from "./recurring-activity";
+
+export class ActivityAttendance extends Entity {
+  static create(from: Date, events: Note[] = []) {
+    const instance = new ActivityAttendance();
+    instance.periodFrom = from;
+    instance.events = events;
+    return instance;
+  }
+
+  periodFrom: Date;
+  periodTo: Date;
+
+  events: Note[] = [];
+
+  activity: RecurringActivity;
+
+  /**
+   * childId to be used as default when getting statistics through the instance's methods (e.g. getEventsPresent)
+   *
+   * if `focusedChild` is set and a method like `getEventsPresent()` is called without parameter
+   * the focusedChild is used as parameter implicitly to simplify calling these methods in templates.
+   */
+  focusedChild: string;
+
+  getEventsTotal(): number {
+    return this.events.length;
+  }
+
+  getEventsWithStatus(
+    status: AttendanceStatus,
+    childId: string = this.focusedChild
+  ): number {
+    return this.events.reduce(
+      (prev: number, currentEvent: Note) =>
+        currentEvent.getAttendance(childId)?.status === status
+          ? prev + 1
+          : prev,
+      0
+    );
+  }
+
+  getEventsPresent(childId: string = this.focusedChild): number {
+    if (!childId) {
+      return this.getEventsPresentAverage();
+    }
+
+    return this.countIndividual(childId, AttendanceCounting.PRESENT);
+  }
+
+  getEventsAbsent(childId: string = this.focusedChild): number {
+    if (!childId) {
+      return this.getEventsAbsentAverage();
+    }
+
+    return this.countIndividual(childId, AttendanceCounting.ABSENT);
+  }
+
+  getAttendancePercentage(childId: string = this.focusedChild) {
+    const present = this.getEventsPresent(childId);
+    const absent = this.getEventsAbsent(childId);
+
+    return present / (present + absent);
+  }
+
+  getAttendancePercentageAverage() {
+    // TODO calculate overall averaged attendance percentage
+    return -1;
+  }
+
+  getEventsPresentAverage() {
+    return this.countAverage(AttendanceCounting.PRESENT);
+  }
+
+  getEventsAbsentAverage() {
+    return this.countAverage(AttendanceCounting.ABSENT);
+  }
+
+  private countIndividual(childId: string, countingType: AttendanceCounting) {
+    return this.events.reduce(
+      (prev: number, currentEvent: Note) =>
+        getAttendanceType(currentEvent.getAttendance(childId)?.status)
```

Comment:
good idea :+1:

Reply:
although filter+length is less optimized for performance, probably: it will create an array containing references to each element ( memory O(n) ) rather than just loading one element, increasing the counter and going to the next ( memory O(1) )